src/plugins/console/public/lib/osd/osd.ts

```diff
@@ -36,69 +38,102 @@ import {
   ListComponent,
   TemplateAutocompleteComponent,
   UsernameAutocompleteComponent,
+  SharedComponent,
 } from '../autocomplete/components';
 
-import $ from 'jquery';
-import _ from 'lodash';
-
-import Api from './api';
+import { Api } from './api';
+
+export type IdAutocompleteComponentFactory = (
+  name: string,
+  parent: SharedComponent,
+  multiValued?: boolean
+) => IdAutocompleteComponent;
+
+export type ComponentFactory = (
+  name: string,
+  parent: SharedComponent | null,
+  multiValued?: boolean | unknown
+) => SharedComponent;
+
+export interface ParametrizedComponentFactories {
+  getComponent: (
+    name: string,
+    parent?: SharedComponent | boolean,
+    provideDefault?: boolean
+  ) => ComponentFactory;
+  index?: (name: string, parent: ListComponent) => IndexAutocompleteComponent | undefined;
+  indices?: (name: string, parent: ListComponent) => IndexAutocompleteComponent | undefined;
+  type?: (name: string, parent: ListComponent) => TypeAutocompleteComponent;
+  types?: (name: string, parent: ListComponent) => TypeAutocompleteComponent;
+  id?: (name: string, parent: SharedComponent) => IdAutocompleteComponent;
+  transform_id?: (name: string, parent: SharedComponent) => IdAutocompleteComponent;
+  username?: (name: string, parent: ListComponent) => UsernameAutocompleteComponent;
+  user?: (name: string, parent: ListComponent) => UsernameAutocompleteComponent;
+  template?: (name: string, parent: ListComponent) => TemplateAutocompleteComponent;
+  task_id?: (name: string, parent: SharedComponent) => IdAutocompleteComponent;
+  ids?: (name: string, parent: SharedComponent) => IdAutocompleteComponent;
+  fields?: (name: string, parent: ListComponent) => FieldAutocompleteComponent;
+  field?: (name: string, parent: ListComponent) => FieldAutocompleteComponent;
+  nodes?: (name: string, parent: SharedComponent) => ListComponent;
+  node?: (name: string, parent: SharedComponent) => ListComponent;
+}
 
 let ACTIVE_API = new Api();
-const isNotAnIndexName = (name) => name[0] === '_' && name !== '_all';
+const isNotAnIndexName = (name: string) => name[0] === '_' && name !== '_all';
 
-const idAutocompleteComponentFactory = (name, parent) => {
+const idAutocompleteComponentFactory: IdAutocompleteComponentFactory = (name, parent) => {
   return new IdAutocompleteComponent(name, parent);
 };
-const parametrizedComponentFactories = {
-  getComponent: function (name, parent, provideDefault) {
-    if (this[name]) {
-      return this[name];
+const parametrizedComponentFactories: ParametrizedComponentFactories = {
+  getComponent(name, parent, provideDefault) {
+    if ((this as any)[name]) {
```

Comment:
`any` type seems not ideal. this is binding to ParametrizedComponentFactories context right?
should we do `getComponent(name: keyof ParametrizedComponentFactories ... )`?

Reply:
I agree, I will try updating it after other 2 related PRs will be merged so that I could transfer the interface to the separate file and update it there.

Reply:
I have updated this part and removed usage of `any`